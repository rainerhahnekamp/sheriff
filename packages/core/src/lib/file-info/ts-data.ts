import type * as ts from 'typescript';
import { FsPath } from './fs-path';
import { TsConfigContext } from "./get-ts-config-context";

export type TsPaths = Record<string, FsPath>;

/**
 * Contains data needed by `traverseFilesystem`.
 */
export type TsData = TsConfigContext & {
  configObject: ReturnType<typeof ts.parseJsonConfigFileContent>;
  cwd: string;
  sys: typeof ts.sys;
}
