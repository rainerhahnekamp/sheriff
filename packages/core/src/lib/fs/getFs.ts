import defaultFs from './default-fs';
import virtualFs from './virtual-fs';
import { Fs } from './fs';

let fsImplementation: 'default' | 'virtual' = 'default';

export function useDefaultFs() {
  fsImplementation = 'default';
  return defaultFs
}

export function useVirtualFs() {
  fsImplementation = 'virtual'
  return virtualFs;
}

const getFs = (): Fs =>
  fsImplementation === 'default' ? defaultFs : virtualFs;

export const isFsVirtualised = () => fsImplementation === 'virtual';

export default getFs;
