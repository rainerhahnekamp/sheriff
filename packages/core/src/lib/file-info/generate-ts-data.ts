import getFs, { isFsVirtualised } from '../fs/getFs';
import * as ts from 'typescript';
import { TsData } from './ts-data';
import { getTsPathsAndRootDir } from './get-ts-paths-and-root-dir';
import { FsPath, toFsPath } from './fs-path';

/**
 * Generates a parsed TypeScript configuration from a given
 * path. The `paths` property will have its values merged
 * with the `baseUrl`.
 *
 * Example:
 *
 * ```json
 * {
 *   baseUrl: './src',
 *   paths: {
 *     'app/*': './app'
 *   }
 * }
 * ```
 *
 * This will return a paths property of `{'app/*': './src/app'}`
 */
export const generateTsData = (tsConfigPath: FsPath): TsData => {
  const { paths, rootDir } = getTsPathsAndRootDir(tsConfigPath);

  const fs = getFs();
  const cwd = fs.getParent(tsConfigPath);
  const configRawContent = getFs().readFile(tsConfigPath);
  const configContent = ts.readConfigFile(tsConfigPath, () => configRawContent);

  const configObject = ts.parseJsonConfigFileContent(
    configContent,
    ts.sys,
    cwd,
  );

  const sys = getTsSys();

  return { paths, configObject, cwd, sys, rootDir };
};

function getTsSys(): ts.System {
  if (isFsVirtualised()) {
    const fs = getFs();
    return {
      fileExists: (path: string) => fs.exists(path),
      readFile(path: string): string | undefined {
        return fs.readFile(toFsPath(path));
      },
    } as unknown as typeof ts.sys;
  } else {
    return ts.sys;
  }
}
