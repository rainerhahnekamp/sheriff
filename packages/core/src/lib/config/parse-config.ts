import { FsPath } from '../file-info/fs-path';
import * as ts from 'typescript';
import { UserSheriffConfig } from './user-sheriff-config';
import getFs from '../fs/getFs';
import { Configuration } from './configuration';
import {
  MissingModulesWithoutAutoTaggingError,
  TaggingAndModulesError,
} from '../error/user-error';
import { defaultConfig } from './default-config';

export const parseConfig = (configFile: FsPath): Configuration => {
  const tsCode = getFs().readFile(configFile);

  const { outputText } = ts.transpileModule(tsCode, {
    compilerOptions: { module: ts.ModuleKind.NodeNext },
  });

  const userSheriffConfig = eval(outputText) as UserSheriffConfig;
  if (userSheriffConfig.autoTagging === false && !userSheriffConfig.tagging) {
    throw new MissingModulesWithoutAutoTaggingError();
  }

  if (userSheriffConfig.tagging && userSheriffConfig.modules) {
    throw new TaggingAndModulesError();
  }

  if (userSheriffConfig.tagging) {
    const {tagging, ...rest} = userSheriffConfig;
    return { ...defaultConfig, ...rest, modules: tagging };

  } else {
    return { ...defaultConfig, ...userSheriffConfig };
  }
};
