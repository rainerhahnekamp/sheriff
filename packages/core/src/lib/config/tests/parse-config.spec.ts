import { describe, it, expect, beforeAll, beforeEach } from 'vitest';
import * as ts from 'typescript';
import { parseConfig } from '../parse-config';
import { toFsPath } from '../../file-info/fs-path';
import getFs, { useVirtualFs } from '../../fs/getFs';
import { MissingTaggingWithoutAutoTagging } from '../../error/user-error';
import '../../test/expect.extensions';

describe('parse Config', () => {
  it('should read value', () => {
    const source = 'export const a = 1';

    const { outputText } = ts.transpileModule(source, {
      compilerOptions: { module: ts.ModuleKind.NodeNext },
    });

    expect(outputText).toMatchSnapshot();
  });

  it('should the sheriff config', () => {
    const tsCode = parseConfig(
      toFsPath(__dirname + '/../../test/sheriff.config.ts'),
    );
    expect(Object.keys(tsCode)).toEqual([
      'version',
      'autoTagging',
      'tagging',
      'depRules',
      'excludeRoot',
      'enableBarrelLess',
      'encapsulatedFolderNameForBarrelLess',
      'showWarningOnBarrelCollision',
      'log',
      'entryFile',
      'isConfigFileMissing',
      'barrelFileName',
    ]);
  });

  describe('virtual fs', () => {
    beforeAll(() => {
      useVirtualFs();
    });

    beforeEach(() => {
      getFs().reset();
    });

    it('should set default values', () => {
      getFs().writeFile(
        'sheriff.config.ts',
        `
import { SheriffConfig } from '@softarc/sheriff-core';

export const config: SheriffConfig = {
  depRules: {
    'noTag': 'noTag',
  },
};
      `,
      );

      const config = parseConfig(
        toFsPath(getFs().cwd() + '/sheriff.config.ts'),
      );
      expect(config).toEqual({
        version: 1,
        autoTagging: true,
        tagging: {},
        depRules: { noTag: 'noTag' },
        enableBarrelLess: false,
        encapsulatedFolderNameForBarrelLess: 'internal',
        showWarningOnBarrelCollision: true,
        excludeRoot: false,
        log: false,
        isConfigFileMissing: false,
        entryFile: '',
        barrelFileName: 'index.ts',
      });
    });

    it('should throw if tagging is missing and autoTagging is disabled', () => {
      getFs().writeFile(
        'sheriff.config.ts',
        `
import { SheriffConfig } from '@softarc/sheriff-core';

export const config: SheriffConfig = {
  autoTagging: false,
};
      `,
      );

      expect(() =>
        parseConfig(toFsPath(getFs().cwd() + '/sheriff.config.ts')),
      ).toThrowUserError(new MissingTaggingWithoutAutoTagging());
    });
  });
});
