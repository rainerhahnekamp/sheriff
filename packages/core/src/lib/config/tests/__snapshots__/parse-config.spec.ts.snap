// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse Config > should read value 1`] = `
""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.a = void 0;
exports.a = 1;
"
`;
